client/test/e2e.test.js

```diff
@@ -0,0 +1,112 @@
+import test from 'tape';
+import puppeteer from 'puppeteer';
+
+function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+const width = 1920;
+const height = 1080;
+
+function browser(width, height) {
+    return puppeteer.launch({
+    headless: true,
+    args: [`--window-size=${width},${height}`],
+    // slowMo: 20 // For testing
+    });
+}
+
+test('e2e-client-test-valid-actions', async (t) => {
+    // Arrange: Set up the page
+    const b = await browser(width, height);
+    const p = await b.newPage();
+    await p.setViewport({ width, height });
+    await p.goto("http://127.0.0.1:8080/");
+    
+    // Act: Test creating a workspace
+    await p.waitForSelector("#add-workspace");
+    await p.click("#add-workspace")
+    await p.waitForSelector("#workspace-name");
+    await p.focus('#workspace-name')
+    await p.keyboard.type('puppeteer')
+    await p.click("#create-workspace")
```

Comment:
This block (and others like it) could use some spacing/formatting to indicate the separate actions being performed. In fact, I would go one further and say that definite, semantic actions like "create a workspace" should be pulled out into a function at the top of the testing module, so there isn't so much repetition of the individual actions throughout the tests.

Reply:
Fixed in 9f65f3d

Reply:
9f65f3d